ci/L0_additional_outputs_vllm/additional_outputs_test.py

```diff
@@ -0,0 +1,185 @@
+        assert len(text_output) > 0, "output is empty"
+        assert text_output.count(" ") > 4, "output is not a sentence"
+
+    def _assert_finish_reason(self, output_finish_reason):
+        for i in range(len(self._responses)):
+            result, error = self._responses[i]["result"], self._responses[i]["error"]
+            assert error is None
+            finish_reason_np = result.as_numpy(name="finish_reason")
+            if output_finish_reason is None or output_finish_reason == False:
+                assert finish_reason_np is None
+                continue
+            finish_reason = finish_reason_np[0].decode("utf-8")
+            if i < len(self._responses) - 1:
+                assert finish_reason == "None"
+            else:
+                assert finish_reason == "length"
+
+    def _assert_cumulative_logprob(self, output_cumulative_logprob):
+        prev_cumulative_logprob = 0.0
+        for response in self._responses:
+            result, error = response["result"], response["error"]
+            assert error is None
+            cumulative_logprob_np = result.as_numpy(name="cumulative_logprob")
+            if output_cumulative_logprob is None or output_cumulative_logprob == False:
+                assert cumulative_logprob_np is None
+                continue
+            cumulative_logprob = cumulative_logprob_np[0].astype(float)
+            assert cumulative_logprob != prev_cumulative_logprob
+            prev_cumulative_logprob = cumulative_logprob
+
+    def _assert_num_token_ids(self, output_num_token_ids):
+        for response in self._responses:
+            result, error = response["result"], response["error"]
+            assert error is None
+            num_token_ids_np = result.as_numpy(name="num_token_ids")
+            if output_num_token_ids is None or output_num_token_ids == False:
+                assert num_token_ids_np is None
+                continue
+            num_token_ids = num_token_ids_np[0].astype(int)
+            # TODO: vLLM may return token ids identical to the previous one when
+            #       streaming, for example:
+            #
+            #       prev: None
+            #       curr: text=' the', token_ids=array('l', [5])
+            #
+            #       prev: text=' the', token_ids=array('l', [5, 1385])
+            #       curr: text=' the term', token_ids=array('l', [5, 1385])
+            #
+            #       prev: text=' the term', token_ids=array('l', [5, 1385, 44])
+            #       curr: text=' the term', token_ids=array('l', [5, 1385, 44])
+            #
+            #       prev: text=' the term', token_ids=array('l', [5, 1385, 44, 48])
+            #       curr: text=' the term “', token_ids=array('l', [5, 1385, 44, 48])
+            #
+            #       If this is no longer the case in a future release, change the assert
+            #       to assert num_token_ids > 0.
```

Comment:
I'm not totally following this section. Can you show me an example output from calling `/generate_stream` on a vllm model with `output_num_token_ids=True` and `stream=True`?

Reply:
On 0.5.3.post1:
```
         text: ' the'
num_token_ids: 1
-----
         text: ' term'
num_token_ids: 1
-----
         text: ''
num_token_ids: 1
-----
         text: ' “'
num_token_ids: 1
-----
         text: 'p'
num_token_ids: 1
-----
         text: 'olar'
num_token_ids: 1
-----
         text: 'ized'
num_token_ids: 1
-----
         text: ''
num_token_ids: 1
-----
         text: '”'
num_token_ids: 1
-----
         text: ' is'
num_token_ids: 1
-----
         text: ' used'
num_token_ids: 1
-----
         text: ' to'
num_token_ids: 1
-----
         text: ' refer'
num_token_ids: 1
-----
         text: ' to'
num_token_ids: 1
-----
         text: ' a'
num_token_ids: 1
-----
         text: ' polar'
num_token_ids: 1
-----
```
which is expected, but on 0.5.5:
```
         text: ' the'
num_token_ids: 1
-----
         text: ' term'
num_token_ids: 0
-----
         text: ''
num_token_ids: 0
-----
         text: ' “'
num_token_ids: 0
-----
         text: 'p'
num_token_ids: 0
-----
         text: 'olar'
num_token_ids: 0
-----
         text: 'ized'
num_token_ids: 0
-----
         text: ''
num_token_ids: 0
-----
         text: '”'
num_token_ids: 0
-----
         text: ' is'
num_token_ids: 0
-----
         text: ' used'
num_token_ids: 0
-----
         text: ' to'
num_token_ids: 0
-----
         text: ' refer'
num_token_ids: 0
-----
         text: ' to'
num_token_ids: 0
-----
         text: ' a'
num_token_ids: 0
-----
         text: ' polar'
num_token_ids: 0
```

Reply:
It appears the previous output (the `token_ids` field) is overwritten by the current output, when the engine is streaming outputs for a request.

Reply:
The issue appears fixed on a later release, i.e. 0.6.3.post1:
```
         text: ' the'
num_token_ids: 1
-----
         text: ' term'
num_token_ids: 1
-----
         text: ''
num_token_ids: 1
-----
         text: ' “'
num_token_ids: 1
-----
         text: 'p'
num_token_ids: 1
-----
         text: 'olar'
num_token_ids: 1
-----
         text: 'ized'
num_token_ids: 1
-----
         text: ''
num_token_ids: 1
-----
         text: '”'
num_token_ids: 1
-----
         text: ' is'
num_token_ids: 1
-----
         text: ' used'
num_token_ids: 1
-----
         text: ' to'
num_token_ids: 1
-----
         text: ' refer'
num_token_ids: 1
-----
         text: ' to'
num_token_ids: 1
-----
         text: ' a'
num_token_ids: 1
-----
         text: ' polar'
num_token_ids: 1
```

Reply:
Out of curiosity is `num_token_ids` actually the requested feature? Or returing the actual `token_ids`? I feel like `num_token_ids` is objectively less useful than returning the real `token_ids`, because you can just do `len(token_ids)` afterwards if needed.

Reply:
ex: https://github.com/triton-inference-server/vllm_backend/pull/41

Reply:
The original ask is for "number of tokens", but I agree that we could return `token_ids` instead, because from a feature perspective we are making the "return_*" space less crowded if the outputting `token_ids` would be requested in the future, in addition to `num_token_ids`.

Reply:
Sorry edited my responses above. While returning `token_ids` is probably more generalized, for large sequences it could actually be more costly than returning a single number if all they want is the count, so I removed that part.

Reply:
[Return token ids instead of number of token ids](https://github.com/triton-inference-server/vllm_backend/pull/70/commits/457eeaa6f23fb480ba38a84b1426e4d67879e58a)

The model.py on the CI pipeline has been replaced with the version on this commit.

Reply:
Yes, we can just add `token_ids` if requested later. Will update the name `num_token_ids` to `num_output_tokens`.

Reply:
[Rename num_token_ids to num_output_tokens](https://github.com/triton-inference-server/vllm_backend/pull/70/commits/dae3c132b25541f82ffe37e5756d9cfb72c4fa22)

The model.py on the CI pipeline has been replaced with the version on this commit.